From 31496e7aca7169e87378f94c1c7119ecfdd8930a Mon Sep 17 00:00:00 2001
From: Will Smillie <will@smillie.me>
Date: Thu, 28 Aug 2025 20:08:15 -0400
Subject: [PATCH] [xhamster] Fix 404 errors on video downloads (Fixes #14145)

- Add proper base64+XOR decryption check via `try_call(lambda:)`
- Update test case metadata
- Remove unnecessary code comments
---
 yt_dlp/extractor/xhamster.py | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/yt_dlp/extractor/xhamster.py b/yt_dlp/extractor/xhamster.py
index 9d90f181ba33..23764708d6dc 100644
--- a/yt_dlp/extractor/xhamster.py
+++ b/yt_dlp/extractor/xhamster.py
@@ -1,3 +1,4 @@
+import base64
 import itertools
 import re
 
@@ -12,6 +13,7 @@
     int_or_none,
     parse_duration,
     str_or_none,
+    try_call,
     try_get,
     unified_strdate,
     url_or_none,
@@ -229,6 +253,13 @@ def get_height(s):
                                 standard_url = standard_format.get(standard_format_key)
                                 if not standard_url:
                                     continue
+                                if standard_url.startswith('eG9yXxAcQ0'):  # base64 of 'xor'
+                                    decoded = try_call(lambda: base64.b64decode(standard_url))
+                                    if decoded and decoded[:4] == b'xor_':
+                                        xor_url_content = decoded[4:]
+                                        standard_url = ''
+                                        for t in range(len(xor_url_content)):
+                                            standard_url += chr(xor_url_content[t] ^ b'xh7999'[t % 6])
                                 standard_url = urljoin(url, standard_url)
                                 if not standard_url or standard_url in format_urls:
                                     continue
